File: python/strategy.py
"""strategy.py — EditStrategy（多 guide 策略）几何评估（批次 E′）。

═══════════════════════════════════════════════════════════════════════════
为什么要有这一层（GPT 裁决 #6/#15/#16）
═══════════════════════════════════════════════════════════════════════════
多 guide 不能只是 `Guide[]`：**双 guide 的风险不是两个单 guide 风险的简单相加**。
- 几何层：两条 guide 的切点共同决定「删掉哪一段」→ 预测连接点（junction）、缺失长度、是否移码。
- 事件层：两条 guide 各自的脱靶位点**两两组合**可能产生额外缺失/易位（pairwise_event），
  不能只报 sum(offtarget(g1), offtarget(g2))。
- 证据层：本模块只做**几何与组合事实**；效率（缺失能否高效发生）不预测——
  文献里缺失效率与长度/位点强相关，本插件不内置该模型。

判定纪律：
  ① 只报可复算的几何事实（切点、区间、junction、长度、移码、PAM 朝向、overlap）；
  ② 未实现的策略（prime_edit / hdr / multiplex…）显式 `implemented: false`，**不许假装支持**；
  ③ pairwise 只在两条 guide 都带 offtarget 命中数据时才算；数据缺失报 `not_searched`（负证据语义）。
"""
from __future__ import annotations

STRATEGIES: dict[str, dict] = {
    'single_cut': {
        'implemented': True,
        'zh': '单切口（nuclease 敲除/敲入起点）',
        'requires': ['candidates'],
        'note': '单 guide 场景直接用 graft_design / graft_rank，不需本工具排策略。',
    },
    'deletion_pair': {
        'implemented': True,
        'zh': '双 guide 缺失（两条 guide 的切点夹住目标区段 → 释放该片段）',
        'requires': ['candidates'],
        'optional': ['region_start_0', 'region_end_0', 'cds_start_0'],
        'note': '切点由候选自带（cut_site_0）；本工具预测缺失区间与连接点。',
    },
    'paired_nickase': {
        'implemented': True,
        'zh': '配对切口酶（两条 nCas9 在相对链上各切一口，降低单切口带来的 indel）',
        'requires': ['candidates'],
        'optional': ['max_offset_bp'],
        'note': '只报几何配对的可行性事实（异链 + 间距），不预测特异性提升幅度。',
    },
    'multiplex_knockout': {
        'implemented': False,
        'zh': '多重敲除（多个独立靶点同时编辑）',
        'requires': ['多个靶区的候选集合'],
        'note': '需要跨靶区的候选分组语义（本版本未实现）——先用 graft_design 分别设计、'
                '各自 graft_rank，再在 EditPlan 里作为多条 run 记录。',
    },
    'base_edit': {
        'implemented': 'elsewhere',
        'zh': '碱基编辑',
        'note': '走 graft_base_edit（窗口几何 + bystander + 密码子后果）。',
    },
    'prime_edit': {
        'implemented': False,
        'zh': 'prime editing（pegRNA）',
        'note': '未实现——如实说明，不假装支持（pegRNA 设计需 PBS/RTT 设计与二级结构评估）。',
    },
    'hdr': {
        'implemented': False,
        'zh': 'HDR 供体设计',
        'note': '未实现——如实说明（供体臂长/同源臂设计需独立的证据与几何规则）。',
    },
}

PAIRWISE_NOTE = (
    'pairwise 分析只回答「两条 guide 的脱靶命中两两组合后，是否存在能共同产生缺失/易位的几何组合」，'
    '**不是**把两条 guide 的风险相加；也不是风险建模（那需要实验与群体数据）。'
)


def _cut_of(c: dict):
    return c.get('cut_site_0')


def _pair_geometry(a: dict, b: dict, *, cds_start_0: int | None = None) -> dict | None:
    """两条候选的几何关系 → 预测缺失区间/junction/移码。切点缺失时返回 None。"""
    ca, cb = _cut_of(a), _cut_of(b)
    if ca is None or cb is None:
        return None
    if ca > cb:
        a, b, ca, cb = b, a, cb, ca
    size = cb - ca
    return {
        'left': {'protospacer': a.get('protospacer'), 'strand': a.get('strand'),
                 'cut_site_0': ca, 'pam': a.get('pam')},
        'right': {'protospacer': b.get('protospacer'), 'strand': b.get('strand'),
                  'cut_site_0': cb, 'pam': b.get('pam')},
        'deleted_interval_0based_half_open': [ca, cb],
        'deleted_size_bp': size,
        'left_flank_end_0based': ca,
        'right_flank_start_0based': cb,
        'junction_rule': '预测连接点 = 左侧序列 [0, cut_left) 直接接右侧序列 [cut_right, len)',
        'protospacers_overlap': bool(
            a.get('start_0') is not None and b.get('start_0') is not None and
            min(a.get('end_0'), b.get('end_0')) > max(a.get('start_0'), b.get('start_0'))),
        'same_strand': a.get('strand') == b.get('strand'),
        'in_frame': (None if cds_start_0 is None else (size % 3 == 0)),
        'frameshift_note': ('缺失长度 %3 != 0 → 移码；%3 == 0 → 保持读码框（对敲除而言移码通常更彻底）'
                            if cds_start_0 is not None else
                            '未声明 cds_start_0：不判断移码（不猜读码框）'),
    }

def _pairwise_offtarget(a: dict, b: dict, *, window_bp: int = 10000, max_events: int = 20) -> dict:
    """两条 guide 的脱靶命中两两组合：同一染色体、间距在窗口内 → 可能共同造成缺失。"""
    ha = (a.get('offtarget_hits') or (a.get('offtarget_summary') or {}).get('hits') or [])
    hb = (b.get('offtarget_hits') or (b.get('offtarget_summary') or {}).get('hits') or [])
    if not ha or not hb:
        missing = []
        if not ha:
            missing.append('left.no_offtarget_hits')
        if not hb:
            missing.append('right.no_offtarget_hits')
        return {'status': 'not_searched', 'missing': missing,
                'note': '两条 guide 都需带 offtarget 命中数据（先跑 graft_offtarget 并把 hits 传进来）；'
                        '缺失即 not_searched，不得当「无风险」'}
    events = []
    for x in ha:
        for y in hb:
            if x.get('chromosome') != y.get('chromosome'):
                continue
            d = abs(int(y.get('position_0based', 0)) - int(x.get('position_0based', 0)))
            if d <= window_bp:
                events.append({
                    'chromosome': x.get('chromosome'),
                    'left_pos_0based': x.get('position_0based'),
                    'right_pos_0based': y.get('position_0based'),
                    'distance_bp': d,
                    'left_mismatches': x.get('mismatches'),
                    'right_mismatches': y.get('mismatches'),
                })
    events.sort(key=lambda e: e['distance_bp'])
    return {'status': 'computed', 'window_bp': window_bp,
            'n_pair_events': len(events), 'events': events[:max_events],
            'note': PAIRWISE_NOTE}


def evaluate_strategy(candidates: list, strategy: str = 'deletion_pair', *,
                      region_start_0: int | None = None, region_end_0: int | None = None,
                      cds_start_0: int | None = None, max_offset_bp: int = 200,
                      pairwise_window_bp: int = 10000, top_n: int = 20) -> dict:
    """评估一个 EditStrategy 的候选组合（几何 + 组合事实；不做效率预测）。"""
    meta = STRATEGIES.get(strategy)
    if meta is None:
        raise ValueError(f'unknown strategy {strategy!r}；可用：{sorted(STRATEGIES)}')
    if meta.get('implemented') is not True:
        return {
            'strategy': strategy, 'implemented': meta.get('implemented'),
            'implementation_state': ('elsewhere' if meta.get('implemented') == 'elsewhere' else 'not_implemented'),
            'zh': meta.get('zh'), 'note': meta.get('note'),
            'verdict': 'cannot_design',
            'honesty': '本插件未实现该策略——不会给出看似可行的假设计；'
                       '如需，请按节点说明改走已实现路径或外部工具。',
            'available_strategies': sorted(STRATEGIES),
        }
    if not isinstance(candidates, list) or len(candidates) < 2:
        raise ValueError('candidates 至少两条（策略评估需要组合）')

    if strategy == 'single_cut':
        return {'strategy': strategy, 'implemented': True, 'verdict': 'use_design_tools',
                'note': meta['note'], 'n_candidates': len(candidates)}

    pairs = []
    for i in range(len(candidates)):
        for j in range(i + 1, len(candidates)):
            g = _pair_geometry(candidates[i], candidates[j], cds_start_0=cds_start_0)
            if g is None:
                continue
            pair = {'pair_id': f'p{len(pairs) + 1}', **g,
                    'offtarget_pairwise': _pairwise_offtarget(
                        candidates[i], candidates[j], window_bp=pairwise_window_bp)}
            if strategy == 'paired_nickase':
                if g['same_strand']:
                    continue
                if not (0 <= g['deleted_size_bp'] <= max_offset_bp):
                    continue
                pair['nickase_offset_bp'] = g['deleted_size_bp']
            elif strategy == 'deletion_pair':
                if region_start_0 is not None and region_end_0 is not None:
                    cl, cr = g['deleted_interval_0based_half_open']
                    # 「覆盖声明区段」= 两个切点**夹住**该区段（缺失区间 ⊇ 区段），
                    # 不是「切点落在区段内」（后者是子区间，反而删不掉区段两端）。
                    # 真实会话实测：旧判据写反，导致工具返回的全是不覆盖区段的组合，
                    # agent 独立枚举后发现「工具评估集 ⊄ 我的覆盖集」。
                    covers = (cl <= region_start_0 and cr >= region_end_0)
                    pair['covers_declared_region'] = covers
                    if covers:
                        pair['extra_deleted_bp_outside_region'] = ((region_start_0 - cl) +
                                                                   (cr - region_end_0))
                    else:
                        pair['reject_reason'] = (
                            f"切点区间 [{cl}, {cr}) 未夹住声明区段 "
                            f"[{region_start_0}, {region_end_0})：两端各需 cut_left ≤ 起点、"
                            f"cut_right ≥ 终点")
            pairs.append(pair)

    n_not_covering = 0
    if strategy == 'deletion_pair':
        n_not_covering = sum(1 for p in pairs if not p.get('covers_declared_region', True))
        pairs = [p for p in pairs if p.get('covers_declared_region', True)]
        # 排序：**区段外附带删除的碱基数越少越精确**（再按缺失长度小→大）
        pairs.sort(key=lambda p: (p.get('extra_deleted_bp_outside_region', 0),
                                  p['deleted_size_bp']))
    else:
        pairs.sort(key=lambda p: p.get('nickase_offset_bp', p['deleted_size_bp']))

    return {
        'strategy': strategy,
        'implemented': True,
        'zh': meta['zh'],
        'n_candidates': len(candidates),
        'n_pairs_evaluated': len(pairs),
        'n_pairs_rejected_not_covering': n_not_covering,
        'pairs': pairs[:int(top_n)],
        'pairs_truncated': len(pairs) > int(top_n),
        'geometry_only': ('本工具只给几何与组合事实：缺失长度/连接点/移码/PAM 朝向/两两脱靶组合。'
                          '缺失效率、配对切口酶的特异性提升幅度等属实验问题，不预测。'),
        'pairwise_rule': PAIRWISE_NOTE,
        'declared_region': ([region_start_0, region_end_0]
                            if region_start_0 is not None and region_end_0 is not None else None),
        'coordinate_note': '所有坐标均为参考序列 + 链的 0-based 半开区间；cut_site_0 来自候选自身。',
    }

File: test/strategy.mjs
// test/strategy.mjs — EditStrategy（多 guide 几何）黄金夹具（人工可算）。
//
// 契约（GPT 裁决 #6/#16）：
//   · 多 guide 不是 Guide[]：几何（切点→缺失区间/junction/移码/PAM 朝向）+ 组合事实（两两脱靶）
//   · **pairwise 不是把两条 guide 的风险相加**；数据缺失 = not_searched（负证据语义）
//   · 未实现的策略（prime_edit / hdr / multiplex…）显式 implemented:false + cannot_design
//     —— 不许给出看似可行的假设计
import { check, summary, op } from './harness.mjs'

// 强一些的伪随机 filler（早期版本用 LCG 低比特 → 退化成 'ACGT' 周期串，夹具失效）
function filler(n, seed) {
  let s = ''
  let x = seed >>> 0
  for (let i = 0; i < n; i += 1) {
    x ^= x << 13; x >>>= 0
    x ^= x >>> 17
    x ^= x << 5; x >>>= 0
    s += 'ACGT'[(x >>> 16) & 3]
  }
  return s
}

// 序列布局（140 nt）：
//   0..32 filler | 33..52 spacer1(+链) | 53..55 PAM1=AGG | 56..83 filler
//   84..86 'CCT' ← 反链 guide 的 + 链 PAM 区（反链 frame 读到 'AGG'）| 87..106 spacer2 | 107..139 filler
// 期望：guide1 '+' cut = 53-3 = 50；guide2 '−' cut = 87+3 = 90
//       → 缺失区间 [50, 90)，长度 40 bp（40 % 3 = 1 → 移码）
const SP1 = filler(20, 111)
const SP2 = filler(20, 222)
const SEQ = filler(33, 333) + SP1 + 'AGG' + filler(28, 444) + 'CCT' + SP2 + filler(33, 555)
const REGION = [50, 90]

const scan = op('guide_enumerate', { sequence: SEQ, editor: 'SpCas9', top_n: 1000 })
const cands = scan.candidates ?? []
const revcomp = (s) => s.split('').reverse().map((c) => ({ A: 'T', C: 'G', G: 'C', T: 'A' }[c])).join('')
const SP2_RC = revcomp(SP2)
// 注意：反链候选的 protospacer 是**反链 frame 的序列**（= +链该区的反向互补），不是 +链序列本身
const g1 = cands.find((c) => c.protospacer === SP1 && c.strand === '+')
const g2 = cands.find((c) => c.protospacer === SP2_RC && c.strand === '-')
check('fixture: both designed guides are found with the expected cut sites',
  g1?.cut_site_0 === 50 && g2?.cut_site_0 === 90,
  `g1=${g1?.cut_site_0} g2=${g2?.cut_site_0} (n_candidates=${cands.length})`)
check('fixture: strands are + and − respectively',
  g1?.strand === '+' && g2?.strand === '-', `${g1?.strand}/${g2?.strand}`)

const CG = { candidates: cands, region_start_0: REGION[0], region_end_0: REGION[1], cds_start_0: 0 }

// ── ① deletion_pair：缺失区间 / 长度 / 移码 ────────────────────────────────
const del = op('evaluate_strategy', { ...CG, strategy: 'deletion_pair' })
const pair = (del.pairs ?? []).find((p) =>
  p.left.protospacer === SP1 && p.right.protospacer === SP2_RC)
check('deletion_pair: the designed pair is predicted with the exact deleted interval',
  pair && JSON.stringify(pair.deleted_interval_0based_half_open) === JSON.stringify(REGION) &&
  pair.deleted_size_bp === 40,
  JSON.stringify(pair && { iv: pair.deleted_interval_0based_half_open, size: pair.deleted_size_bp }))
check('deletion_pair: the pair covers the declared region',
  pair?.covers_declared_region === true, JSON.stringify(pair?.reject_reason))
check('deletion_pair: covering is "cuts flank the region", with the outside-deletion cost reported',
  pair?.extra_deleted_bp_outside_region === 0,
  JSON.stringify({ extra: pair?.extra_deleted_bp_outside_region }))
check('deletion_pair: non-covering pairs are rejected and counted (not silently mixed in)',
  typeof del.n_pairs_rejected_not_covering === 'number' &&
  del.n_pairs_rejected_not_covering >= 1 &&
  (del.pairs ?? []).every((p) => p.covers_declared_region === true),
  JSON.stringify({ rejected: del.n_pairs_rejected_not_covering, returned: (del.pairs ?? []).length }))
check('deletion_pair: pairs are sorted by "least extra deletion outside the region" first',
  (del.pairs ?? []).every((p, i, arr) => i === 0 ||
    (arr[i - 1].extra_deleted_bp_outside_region ?? 0) <= (p.extra_deleted_bp_outside_region ?? 0)),
  JSON.stringify((del.pairs ?? []).slice(0, 5).map((p) => p.extra_deleted_bp_outside_region)))
check('deletion_pair: frameshift computed from the declared CDS (40 % 3 !== 0)',
  pair?.in_frame === false && /移码/.test(String(pair?.frameshift_note)),
  JSON.stringify({ in_frame: pair?.in_frame }))
check('deletion_pair: junction rule is stated (no hidden assumptions)',
  /预测连接点/.test(String(pair?.junction_rule)))
check('deletion_pair: geometry-only discipline is stated',
  /只给几何与组合事实/.test(String(del.geometry_only)) && /不预测/.test(String(del.geometry_only)))

// ── ② pairwise 脱靶：无数据必须 not_searched（不是「无风险」）──────────────
check('pairwise off-target without hit data → not_searched + missing list',
  pair?.offtarget_pairwise?.status === 'not_searched' &&
  (pair.offtarget_pairwise.missing ?? []).length === 2 &&
  /不得当「无风险」/.test(String(pair.offtarget_pairwise.note)),
  JSON.stringify(pair?.offtarget_pairwise))

// ── ③ pairwise 脱靶：给了 hits 才算，且只算同染色体、按距离排序 ────────────
const withHits = cands.map((c) => {
  if (c.protospacer === SP1) {
    return { ...c, offtarget_hits: [
      { chromosome: 'chrX', position_0based: 1000, mismatches: 2 },
      { chromosome: 'chrY', position_0based: 5, mismatches: 3 },       // 不同染色体 → 不计
    ] }
  }
  if (c.protospacer === SP2_RC) {
    return { ...c, offtarget_hits: [
      { chromosome: 'chrX', position_0based: 1850, mismatches: 3 },    // 同染色、距 850
      { chromosome: 'chrX', position_0based: 9300, mismatches: 3 },    // 同染色、距 8300
    ] }
  }
  return c
})
const del2 = op('evaluate_strategy', { ...CG, candidates: withHits, strategy: 'deletion_pair' })
const pair2 = (del2.pairs ?? []).find((p) => p.left.protospacer === SP1 && p.right.protospacer === SP2_RC)
const pw = pair2?.offtarget_pairwise
check('pairwise: computed only from same-chromosome pairs, sorted by distance',
  pw?.status === 'computed' && pw.n_pair_events === 2 &&
  pw.events[0].distance_bp === 850 && pw.events[1].distance_bp === 8300,
  JSON.stringify({ status: pw?.status, n: pw?.n_pair_events, ev: (pw?.events ?? []).map((e) => e.distance_bp) }))
check('pairwise: states that it is not risk summation',
  /不是.*相加/.test(String(pw?.note)), String(pw?.note).slice(0, 60))

// ── ④ paired_nickase：只保留异链且间距在窗口内 ────────────────────────────
const nick = op('evaluate_strategy', { ...CG, strategy: 'paired_nickase', max_offset_bp: 100, top_n: 1000 })
const nickPair = (nick.pairs ?? []).find((p) => p.left.protospacer === SP1 && p.right.protospacer === SP2_RC)
check('paired_nickase: opposite-strand pair kept with the nick offset reported',
  nickPair?.nickase_offset_bp === 40, JSON.stringify(nickPair && { off: nickPair.nickase_offset_bp }))
check('paired_nickase: every returned pair is cross-strand',
  (nick.pairs ?? []).every((p) => p.same_strand === false),
  JSON.stringify((nick.pairs ?? []).map((p) => p.same_strand)))
check('paired_nickase: offset window enforced (no pair beyond max_offset_bp)',
  (nick.pairs ?? []).every((p) => p.nickase_offset_bp <= 100))

// ── ⑤ 未实现策略：显式拒绝，不给假设计 ────────────────────────────────────
for (const s of ['prime_edit', 'hdr', 'multiplex_knockout']) {
  const r = op('evaluate_strategy', { ...CG, strategy: s })
  check(`unimplemented strategy "${s}" refuses honestly`,
    r.implemented === false && r.verdict === 'cannot_design' &&
    /未实现/.test(String(r.honesty)),
    JSON.stringify({ impl: r.implemented, verdict: r.verdict }))
}
const elsewhere = op('evaluate_strategy', { ...CG, strategy: 'base_edit' })
check('base_edit strategy points to the dedicated tool instead of faking geometry',
  elsewhere.implementation_state === 'elsewhere' && /graft_base_edit/.test(String(elsewhere.note)),
  JSON.stringify(elsewhere.implementation_state))

let singleCut = null
let singleCutError = ''
try {
  singleCut = op('evaluate_strategy', { candidates: [g1], strategy: 'single_cut' })
} catch (e) {
  singleCutError = e.message
}
check('single_cut accepts one candidate and delegates to the design tools',
  singleCut?.implemented === true &&
  singleCut?.verdict === 'use_design_tools' &&
  singleCut?.n_candidates === 1,
  singleCutError || JSON.stringify(singleCut))

// ── ⑥ 输入卫生 ────────────────────────────────────────────────────────────
let badStrategy = ''
try { op('evaluate_strategy', { ...CG, strategy: 'magic' }) } catch (e) { badStrategy = e.message }
check('unknown strategy fails loudly and lists the known ones',
  /unknown strategy/.test(badStrategy) && /deletion_pair/.test(badStrategy), badStrategy.slice(0, 140))

let tooFew = ''
try { op('evaluate_strategy', { candidates: [g1], strategy: 'deletion_pair' }) } catch (e) { tooFew = e.message }
check('fewer than two candidates fails loudly', /至少两条/.test(tooFew), tooFew.slice(0, 120))

summary('strategy')
